### ISSUE TITLE
Proposal: Allow setting readonly fields to null in Dispose

### ISSUE THREAD

**Issue by JohanLarsson** [dotnet/csharplang#542]
```c#
public class Foo : IDisposable
{
    private readonly IDisposable disposable;

    public Foo()
    {
        this.disposable = new Disposable();
    }

    public void Dispose()
    {
        this.disposable.Dispose();
        this.disposable = null;
    }
}
```

**Comment by HaloFour** [comment]
The CLR forbids assigning `readonly`/`initonly` fields anywhere outside of the constructor.  The CLR has no special knowledge of `IDisposable` or the `Dispose` method.  As a language C# couldn't do this without modifications to the CLR.

**Comment by jnm2** [comment]
You could do some wacky things to protected readonly fields by inheriting and implementing `IDisposable`.

**Comment by DavidArno** [comment]
If you want to modify a field, then don't make it `readonly`. Disposable items are inherently mutable, so I'd consider:
```cs
private readonly IDisposable disposable;
```
as a code smell and worthy of a compiler/analyzer warning that `IDisposable` fields should not be marked read-only.

Just get rid of that `readonly` and your `Dispose` code will work just fine.

**Comment by JohanLarsson** [comment]
@DavidArno of course we can get rid of the readonly and make the code compile, problem is that it makes the code less clear, `readonly` is nice as it makes it obvious that there is no mutation going on.
The dispose method is special though as it is normally the last time an instance is touched from the outside. After the instance is disposed touching it should normally throw an `ObjectDisposedException`. Yes there are exceptions like `Task<T>`.

**Comment by bondsbw** [comment]
If you want to prevent use of `disposable` after calling `Dispose()`, then (with this proposal) you would need null checks for each use of the field.

But is that better than checking a flag called `IsDisposed`?  That would communicate purpose better.

**Comment by jnm2** [comment]
@DavidArno Thoroughly disagree that readonly _references_ equate to immutable _graphs_. Whether or not you specify `readonly` is independent of whether the object it is referencing at any given time can mutate.

**Comment by JohanLarsson** [comment]
@bondsbw agreed a flag is clearer. 

Most of the time I don't set things to null in the dispose method so `readonly` is not an issue, was still curious about what others think so I posted this issue.

**Comment by yaakov-h** [comment]
I usually don't bother nulling out the `IDisposable` object either, since the documentation for `IDisposable.Dispose()` states:

> If an object's Dispose method is called more than once, the object must ignore all calls after the first one. The object must not throw an exception if its Dispose method is called multiple times.

I don't think I've yet encountered an object that visibly violates this.

@JohanLarsson `readonly` makes it nice as it makes it obvious (and enforces) that there is no mutation going on, however by nulling out the reference there **is** mutation going on. You can't really have it both ways.

**Comment by JohanLarsson** [comment]
@yaakov-h yeah it would be mutation but it would be restricted to ctor & dispose.

**Comment by yaakov-h** [comment]
that's not `readonly` then, that's `readonlyexceptonespecialmethod`.

**Comment by DavidArno** [comment]
@HaloFour,

I'm confused by your assertion regarding the CLR and read-only fields. Unless reflection uses a weird loophole, of course, as the latter approach can be used to modify read-only fields:
```cs
using static System.Console;
using static System.Reflection.BindingFlags;

namespace MutableReadonly
{
    internal class Program
    {
        private readonly string _foo = "foo";

        private static void Main()
        {
            var x = new Program();
            WriteLine(x._foo); // prints foo
            typeof(Program).GetField("_foo", Instance|NonPublic).SetValue(x, "bar");
            WriteLine(x._foo); // prints bar
        }
    }
}
``` 

@jnm2,

To be clear, I'm not claiming that `readonly` makes an instance immutable. I'm unclear whether you thought I was saying that, though.

To my mind, the FxCop violation, [CA2104: Do not declare read only mutable reference types](https://msdn.microsoft.com/en-us/library/ms182302.aspx) hits the nail on the head. It's a shame in my view that this check was buried away in a rarely used tool, rather than being a compiler warning, but it is what it is. 

If I see `readonly`, I want to be able to assume that the thing in question really is read-only, and ideally that it is immutable too. Therefore, in my view, the following code is a misuse of `readonly`:
```cs
private readonly SomeType _field;

...

_field.X = 1;
```
I'd hesitate to call in an anti-pattern, but I definitely view it as poorly written code. You are of course entitled to disagree as that is just my opinion.

**Comment by JohanLarsson** [comment]
@DavidArno here is a bit of trivia for you:
```c#
struct Foo
{
    public readonly int Value;

    public Foo(int value)
    {
        this.Value = value;
    }

    public void Update(int newValue)
    {
        this = new Foo(newValue);
    }
}
```

**Comment by DavidArno** [comment]
@JohanLarsson,

😱😱😱😱😱

Why is that even possible? That's not an anti-pattern; that's a war crime! 😀

**Comment by JohanLarsson** [comment]
@yaakov-h a pedant would say *twospexialmethods(ctor & dispose)

**Comment by jnm2** [comment]
@DavidArno I'm surprised that such a central tenet of C# is considered poor practice. It's still valuable to prevent mutation of the reference even when the target mutates. It's so common in OOP where the central  tenet is that every object is a black box and it should be irrelevant whether it has internal state.

**Comment by jnm2** [comment]
@DavidArno As I've said elsewhere, going outside the language and type system whether via reflection or via unmanaged pointers to managed memory means leaving safety behind. You're left with memory which is inherently mutable; thus, disallowing setting a field in reflection would just be an invitation for a library that circumvents the check.

**Comment by DavidArno** [comment]
@jnm2,

Reflection and unmanaged pointers are orthogonal, surely? Reflection is only permitted with managed data, as far as I know.

But that's not the point I was trying to make. @HaloFour's statement that "*The CLR forbids assigning `readonly/initonly` fields anywhere outside of the constructor*" seems plain wrong to me as reflection is a function of the CLR and so the CLR does allow re-assigning `readonly/initonly` fields under certain circumstances. Or have I misunderstood the relationship between the CLR and reflection?

**Comment by DavidArno** [comment]
@jnm2,
> I'm surprised that such a central tenet of C# is considered poor practice. It's still valuable to prevent mutation of the reference even when the target mutates.

I disagree. That concept of "value" is really just complacency. It feels like it's offering safety, but can mislead one into thinking things are safer than they are. So in my view at least, one is better off making it clear that that target can mutate by not using `readonly`.

> It's so common in OOP where the central tenet is that every object is a black box and it should be irrelevant whether it has internal state.

That is a flawed central tenet though in this modern world of things needing to be thread-safe. I have a genuine need to know if that box can mutate its state, for the rules change around how it can be safely used if it does mutate.

**Comment by HaloFour** [comment]
@DavidArno 

Reflection has always been an unfortunately backdoor to that restriction.  I'm not sure why the CLR permits it.  Using reflection does immediately bump the assembly to require full trust.

Anywho, I don't find the fact that reflection can bypass the restriction as reason enough to ignore the restriction.

**Comment by DavidArno** [comment]
@HaloFour,

Thanks for the clarification. 100% agree with "*I don't find the fact that reflection can bypass the restriction as reason enough to ignore the restriction*".

**Comment by jnm2** [comment]
> I have a genuine need to know if that box can mutate its state

In C#, given its background and the majority of code, nothing is assumed to be threadsafe unless documented as such. Thread safety typically represents overhead and what's more, can typically be retrofitted when necessary. There are exceptions of course, but that's been the rule.

Guessing about most people but speaking for myself I know precisely what `readonly` means and does not mean, so complacency is hardly in the picture. No different than `const` in JavaScript or C++. It's a very precise and familiar concept.

**Comment by CyrusNajmabadi** [comment]
> Why is that even possible? 

Because its' a value.  And values are mutable.   I mean, do you think the following should be allowed:


```c#
Struct s = ...
Struct t = ...
s = t;
```

?

Inside a call to anything on 's' "this" must of course be a reference to 's'.  If 'this' was not a reference, then methods on structs could not actually mutate anything.  i.e.:

```c#
Struct s = ...
s.Mutate();
```

Wouldn't be possible.

**Comment by DavidArno** [comment]
@CyrusNajmabadi,
> Inside a call to anything on 's' "this" must of course be a reference to 's'. If 'this' was not a reference, then methods on structs could not actually mutate anything

Sounds perfect to me. 😕 

MS advice used to be that structs should be immutable. In my view, changing that to effectively encourage making them as mutable as possible was a very bad decision. But we aren't ever going to agree on that point

**Comment by CyrusNajmabadi** [comment]
> In my view, changing that to effectively encourage making them as mutable as possible was a very bad decision

That has never been the case.

What is the case is that we advise people understand what it means for a struct to be mutable, and to use mutable structs *when* appropriate.

> But we aren't ever going to agree on that point

That's likely very true.  Mutable structs are just too valuable and too important to not have.  For example, without them, enumerating a List would actually force you to allocate, and that would be terrible.

--

Mutable structs are actually amazingly good for some sorts of things.  For example, as we discussed in the Tuple threads, they allow you to get a Tuple implementation that is:

1. Efficient.
2. Safe to share.

Achieving that without mutable structs is amazingly difficult.  And these are two aspects of tuples that people care  *a lot* about.

**Comment by Thaina** [comment]
You can still use anything that already disposed. Dispose is not destructor

**Comment by DavidArno** [comment]
@CyrusNajmabadi 
> Mutable structs are actually amazingly good for some sorts of things. For example, as we discussed in the Tuple threads, they allow you to get a Tuple implementation that is:
> 1. Efficient.
> 2. Safe to share.
>
> Achieving that without mutable structs is amazingly difficult. And these are two aspects of tuples that people care a lot about.

1. Mutating the field of a mutable struct offers a minimal (my experiments suggest around 10% for a three field tuple) speed improvement over creating new copies of a immutable struct. Doing that with tuples means that tiny efficiency comes at the cost of breaking a fundamental principle of what a tuple is supposed to be.
2. Sure it might be "amazingly difficult" to achieve that tiny speed increase without mutable structs. That doesn't mean tuples needed to be mutable. Those needing that tiny speed increase could have used other structs.
3. Some people may well care a lot that their tuples be mutable. Many others cared a lot that you corrupted the basic principle of tuples to satisfy the first group.

You won't ever please all the people .. so don't try to pretend you did. The decision was made to please the "speed heads" over the "software craftsmanship brigade". Obviously I would have preferred the opposite decision, but you have made clear that your biggest customers want speed, not quality, and their opinions matter to you.  

I accept that structs sometimes need to be mutable. I just think it sad that the team have shifted the official guidance from "structs should be immutable unless really necessary" to "make 'em all mutable; it's efficient you know".

**Comment by CyrusNajmabadi** [comment]
> Doing that with tuples means that tiny efficiency comes at the cost of breaking a fundamental principle of what a tuple is supposed to be.

Could you explain what this "fundamental principle of what a tuple is supposed to be" is?

**Comment by CyrusNajmabadi** [comment]
> I just think it sad that the team have shifted the official guidance from "structs should be immutable unless really necessary" to "make 'em all mutable; it's efficient you know".

We have never done that.  Please stop putting forth strawmen.  

As i've pointed out in these conversations in the past, mutable structs are generally a poor idea *because* the copying nature of structs fights against mutations.  i.e. if you have a mutable struct, and you pass it somewhere, then mutations you make to one won't be seen by the other.  For some data types, this is highly undesirable and will likely not be what you want.

However, for tuples, this is not the case.  This behavior is *exactly* what we felt was appropriate.  We wanted tuples to act as values.  If i pass my tuple to you you see that tuple as a value.  i.e. it will be a copy of the tuple i have.  If i then mutate my copy, yours won't chant.  That's *desirable* behavior given how we think about Tuples, and so mutable structs are a great implementation choice because they give us that behavior, with the best perf possible on all the VMs out there.

--

A core problem here @DavidArno is that you've somehow taken the message of "use mutable structs when appropriate" to one of "mutable structs are bad".  Mutable structs are not bad.  Mutable structs just come with a set of behaviors and properties that may make something suitable/unsuitable for a particular domain.

What we do here is look at all our options and weigh the pros/cons of all of them.  In the case of a mutable struct, we don't just go "oh, it's amutable struct, thus it's bad".  We say:

1. what are immutable classes good/bad at?  They're good at being immutable.  So you can pass them around safely.  But they're extremely expensive.  Especially when code wants to actually change values.  Hrmm... that's unfortunate.
2. what are mutable classes good/bad at?  They're great to be able to efficiently mutate data.  But they're terrible for passing around.  now you can't safely give you value to someone else.  This has all sorts of implications for scaling/concurrency.  Also, they're very expensive.  That's unfortunate.
3. What are immutable structs good/bad at?  They're great for passing data around.  But they're not great for mutating data.  That's unfortunate.
4. What are mutable structs good/bad at? They're great for mutating *your own data*.  They're great for passing around.  They're super fast.  They're not great if you want to keep yourself from mutating your own data. 

All these had pros/cons.  We looked at all of these and realized right away that classes were not going to cut it.  So it came down mutable or immutable structs.  We weighed the pros/cons against each other, and we found the perf benefits compelling enough to want to take that, and we found the 'con' of being able to mutate your own data to not be something we felt was an actual problem.  C# has always allowed you to mutate your own data.  in practice we find that preventing mutating is the most valuable when sharing with other modules.  And mutable structs provided for that.

So, we made an informed, reasoned decision on this.  This was informed with decades of years of knowledge from people all over the developer spectrum inside MS and outside.  It was informed by data we collected across tons of platforms, and discussions and investigations we made with many.

We ended up with a decision you don't like.  I get that.  But please, accept what i said before:

> What is the case is that we advise people understand what it means for a struct to be mutable, and to use mutable structs when appropriate.

Again, please stop with your strawmen of "make 'em all mutable; it's efficient you know".  It's not helpful and it is bizarre given that *over and over* again i've been clear about this.

**Comment by DavidArno** [comment]
@CyrusNajmabadi,

I had deleted that comment as I decided that it wasn't worth going back over the same old arguments. I forgot though that you prefer to handle comments via email, so deleting didn't help ... so as you responded, I might as well...

> 1. what are immutable classes good/bad at? They're good at being immutable. So you can pass them around safely. But they're extremely expensive. Especially when code wants to actually change values. Hrmm... that's unfortunate.

How is it expensive to pass immutable classes around? They are passed by reference: they are very cheap. The real downsides to immutable classes is that `null` creeps in and makes life hard and they are slower to allocate than structs.

> 2. what are mutable classes good/bad at? They're great to be able to efficiently mutate data. But they're terrible for passing around. now you can't safely give you value to someone else. This has all sorts of implications for scaling/concurrency. Also, they're very expensive. That's unfortunate.

"But they're terrible for passing around. now you can't safely give you value to someone else" makes no sense. The whole reason for making them mutable is to allow them to be mutated. If you don't want them to be mutated, then don't make them mutable. Again how are they expensive to pass around?

> 3.  What are immutable structs good/bad at? They're great for passing data around. But they're not great for mutating data. That's unfortunate.

No, they are really bad for passing data around. That's why you added ref returns to the language: because coping structs is expensive. Why is it *unfortunate* that they offer protection against mutation?

> 4.  What are mutable structs good/bad at? They're great for mutating your own data. They're great for passing around. They're super fast. They're not great if you want to keep yourself from mutating your own data.

No, they are poor for passing around . That's why you added ref returns to the language: because coping structs is expensive. Structs are much faster than objects when it comes to allocation (which is why mutable structs offer such a tiny speed advantage over immutable structs). They're much slower when it comes to copying them around though. That's why you added ref returns to the language.

**Comment by DavidArno** [comment]
@CyrusNajmabadi 
> Could you explain what this "fundamental principle of what a tuple is supposed to be" is?

Sure. As [Mads says in *Proposal: Language support for Tuples*](https://github.com/dotnet/roslyn/issues/347):
> In their very motivation, though, tuples are ephemeral. You would use them when the parts are more important than the whole. So the common pattern would be to construct, return and immediately deconstruct them. In this situation structs are clearly preferable.

The whole idea of tuples is that the are short-lived values that are constructed, copied once and deconstructed. They come from the "immutable by default" world of FP. Their fundamental principle is that they are ephemeral read-only values. 

Creating a tuple, then modifying it a bit and a bit more, before copying it elsewhere completely misses the point of tuples. You can call my claim that ""make 'em all mutable; it's efficient you know" is a strawman, but I disagree. It's an accurate representation of what you did with tuples: you put speed before "common pattern" good design.

**Comment by Thaina** [comment]
@DavidArno I think the word "Short Live" could be argue how short it is. Some might define short to be on function. Some would be on one stack. For me I would define short as one game loop or one event of the game. People who use linq are using anonymous object as short live, but it still living for 5-6 function cycle

So it could be destructed fast. But it not means it must be immutable

**Comment by DavidArno** [comment]
@Thaina,

No, there isn't really any argument over what "short-lived" means for tuples. Mads defined it well: the lifetime of a tuple "pattern" is to be constructed in one function, returned to the caller and deconstructed.

**Comment by GeirGrusom** [comment]
> They come from the "immutable by default" world of FP.

They come from math. It's an ordered, finite array of values. That's it. It makes no statement about mutability.

Mutability is an advantage for C# because deconstruction can turn into a single `stloc` instruction. It's a clear win.

**Comment by jnm2** [comment]
The core principle of tuples is to be an analog of an argument list, to the extent possible.
Parameters are mutable; therefore, the simplest expectation would be that tuple fields are mutable as well.

**Comment by CyrusNajmabadi** [comment]
> "How is it expensive to pass immutable classes around?"

I didn't say it was expensive to pass them around.  I said "But they're extremely expensive.  Especially when code wants to actually change values. "

Each immutable instance requires an allocation.  Allocations are not free, and we've learned *painfully* over the past decade exactly how non-free they are.  In a world where tuples are immutable classes, then every element change is also an allocation.

Right now Roslyn *itself* can't use several C# features in certain areas because these costs are too high.  I'm spending the majority of my engineering efforts on addressing this in the IDE because allocation cost, and hte impact to the user, is the single largest perf issue we have.

**Comment by CyrusNajmabadi** [comment]
> "But they're terrible for passing around. now you can't safely give you value to someone else" makes no sense. The whole reason for making them mutable is to allow them to be mutated. If you don't want them to be mutated, then don't make them mutable. 

That's why i said it was a *con*.  Because now that you have mutable classes, you can't safely share them.  Someone might accidently mutate the tuple and break your copy.   This is not an issue with the struct-form.

> Again how are they expensive to pass around?

I did not say they are expensive ot pass around.  I said they are expensive just as immutable classes are.  See my previous post on this.  Allocs and GC are extremely detrimental to performance in many domains.  As people want things to run well on everything from Mobile to the Datacenter (where inefficiency translates to wasted money), we don't want to impose unnecessary costs on people.

**Comment by CyrusNajmabadi** [comment]
> No, they are poor for passing around . That's why you added ref returns to the language: because coping structs is expensive.

This is the case when structs are quite large.  The expected usage pattern in practice for tuples is usually <4 elements.  In that case structs are fantastic for passing around.  

*And* if you are using a very large struct, then yes you can use ref.  value-structs allow for fast perf for the overwhleming common case, as well as fast perf for the uncommon case.  Importantly, they are safe to share around.

**Comment by CyrusNajmabadi** [comment]
>> Sure. As Mads says in Proposal: Language support for Tuples:
> In their very motivation, though, tuples are ephemeral. You would use them when the parts are more important than the whole. So the common pattern would be to construct, return and immediately deconstruct them. In this situation structs are clearly preferable.

Nothing about that states that they should be immutable.

> They come from the "immutable by default" world of FP. 

They really don't.  Tuples predate FP *significantly*.

> You can call my claim that ""make 'em all mutable; it's efficient you know" is a strawman, but I disagree. It's an accurate representation of what you did with tuples: you put speed before "common pattern" good design. 

That was not your claim.  Your claim was:

"I just think it sad that the team have shifted the official guidance from "structs should be immutable unless really necessary" to "make 'em all mutable; it's efficient you know"."

We have not shifted our guidance on structs.  Our guidance on structs is that *generally* being mutable is probably not what you want.  However, in some domains it's *exactly* what you want.  And when that is the case, you should absolutely use them as they're great for this purpose.

**Comment by HaloFour** [comment]
@CyrusNajmabadi @DavidArno 

My karma ran over your dogma.  🐕 🚗 

Wait, what was this proposal about again?

**Comment by CyrusNajmabadi** [comment]
> No, there isn't really any argument over what "short-lived" means for tuples. Mads defined it well

I think you're mixing up concepts here.  Take Roslyn as a good example.  The Roslyn API embraces immutability heavily.  Huge swaths of our API give you a surface area that is observably immutable and is safe to use a in massively threaded scenarios.   

However, internally, mutability is embraced heavily to be a means to present this immutable facade.  It's the basis of incremental, lazy computation.  *Something which came heavily from the FP world*.     

We achieve a scalable immutable API *precisely* by embracing mutability and using it effectively to provide something actually efficient.

**Comment by PathogenDavid** [comment]
>HaloFour: The CLR forbids assigning `readonly`/`initonly` fields anywhere outside of the constructor.

>DavidArno: I'm confused by your assertion regarding the CLR and read-only fields. Does reflection use a weird loophole?

@DavidArno

It's maybe worth noting that the runtime will happily JIT code for non-ctor methods which modify `readonly` fields. Unsurprisingly however, it does not produce a verifiable assembly. Basically it works, it just isn't kosher.

Here's a proof of concept I made to test this a while back: [ForceWriteReadonly](https://gist.github.com/PathogenDavid/28c52c6a90fcc4c28a9d66981c92ba5d). I can upload the full solution if someone wants to build this atrocity for some reason. (This also includes an example for calling a constructor as if it were a normal method, which somehow feels even more evil.)

I imagine that producing verifiable assemblies is basically mandatory for Roslyn, so without CLR changes this proposal is a non-starter. IMO, CLR should not be aware of the IDisposable pattern, so I wouldn't think it *should* change either.

----

My primary motivation for wanting a feature similar to this is that we null out `IDisposable` fields so that accidental access of disposed fields causes the offending code to die in a fiery `NullReferenceException`. However, as pointed out earlier in this thread, you should be getting an `ObjectDisposedException` before that happens, so it really only prevents cases where you aren't checking if the object has been disposed.

As shown above, you could use an evil Fody weaver like I've shared to accomplish this feature if you really want it. However, you're better off using something like [Fody.Janitor](https://github.com/Fody/Janitor) to add automatic throwing of `ObjectDisposedException` instead.

TL;DR: You can do this if you really want to using evil IL manipulation, but you probably don't actually need this.

**Comment by GeirGrusom** [comment]
Couldn't code that alters readonly values in principle cause a segfault if the CLR for some reason would place the value in read-only memory? The CLR does already make assumptions that readonly values never change.

Otherwise it's just as @DavidArno says: if you want to change it's value it shouldn't be readonly. What's the point of marking something readonly when it in practice isn't.

**Comment by PathogenDavid** [comment]
>CLR for some reason would place the value in read-only memory

I don't picture a situation where CLR could/would do that. It'd require allocating a whole page of memory (usually 4 KB on Windows) for `readonly` fields. If it wanted to reuse that 4 KB page (because few objects will have 4 KB of `readonly` fields) then it'd also mean changing the protection of that page when a constructor is run, and changing it back when it is finished.

>What's the point of marking something readonly when it in practice isn't.

I'm not really arguing in favor of this proposal, but to play devil's advocate: You might want to change `readonly` fields in the `Dispose` method because the fields should never be read ever again, so you make them an invalid value in case someone does read them by mistake. Essentially `readonly` becomes "readonly during the valid lifetime of the object" instead of "readonly after construction".

**Comment by GeirGrusom** [comment]
> I don't picture a situation where CLR could/would do that.

It's an implementation detail. The point is more that the CLR assumes that `initonly` doesn't change value after creation, and can do whatever it likes to protect that (or do nothing at all). It could potentially pool objects it finds where all fields are `readonly`, or it may not.

I wouldn't bet anything on all implementations of the CLR allowing code that writes to `readonly` fields to execute in the future.

**Comment by jnm2** [comment]
@PathogenDavid Except that from the point of view of the runtime, the object's valid lifetime is from construction to finalization and IDisposable is irrelevant.

**Comment by PathogenDavid** [comment]
>I wouldn't bet anything on all implementations of the CLR allowing code that writes to `readonly` fields to execute in the future.

For sure. I presented my Fody weaver for the sake of trivia for the two of the two bigger implementations of CLR as that exist today. It's super evil and you should **never** do it, I did not mean to imply otherwise.

>Except that from the point of view of the runtime, the object's valid lifetime is from construction to finalization and IDisposable is irrelevant.

Correct, but `IDisposable` is the mechanism for the developer to explicitly say when an object's valid lifetime ends as far as the developer is concerned, which is what I was referring to. This proposal (which I am not in favor of, BTW) would seek to make CLR vaguely ware of `IDisposable` if it were implemented (which it probably won't/shouldn't be.)

**Comment by jnm2** [comment]
I'm also not in favor of moving that direction. https://github.com/dotnet/csharplang/issues/1451#issuecomment-380103599
